File: src/ndcompression/_internals/_nd_shape.py
import itertools
import math


class NDShape:
    def __init__(self, shape: list[int]):
        self.shape = shape
        self.strides = self.calculate_strides(shape)
        self.len = math.prod(self.shape)

    @staticmethod
    def calculate_strides(shape: list[int]) -> list[int]:
        strides = []
        current_stride = 1
        for dim in reversed(shape):
            strides.append(current_stride)
            current_stride *= dim
        return list(reversed(strides))

    def flat_index(self, indexes: list[int]) -> int:
        return sum(index * stride for index, stride in zip(indexes, self.strides))

    def multi_index(self, index: int) -> list[int]:
        remaining = index
        indices = []
        for stride in self.strides:
            indices.append(remaining // stride)
            remaining %= stride
        return indices

    def slices(self):
        rows_template = self.__rows_template()
        for shape_index, length in enumerate(self.shape):
            num_rows = math.prod(self.shape) // length
            for row in itertools.islice(rows_template, num_rows):
                for length_index, indecies in enumerate(row):
                    indecies = list(indecies)
                    indecies[shape_index] = length_index
                    row[length_index] = tuple(indecies)
                yield row

    def __trimmed_shape(self):
        for shape_index in range(len(self.shape)):
            yield [
                1 if i == shape_index else self.shape[i] for i in range(len(self.shape))
            ]

    def __column_parts(self):
        for column in self.__trimmed_shape():
            yield itertools.product(*[range(length) for length in column])

    def __column(self):
        for column in itertools.chain(*self.__column_parts()):
            yield tuple(column)

    def __rows_template(self):
        column = self.__column()
        for length in self.shape:
            for _ in range((math.prod(self.shape) // length)):
                yield list(itertools.repeat(next(column), length))

    def __len__(self):
        return self.len

File: src/nd_shape.rs
use itertools::Itertools;
use std::collections::VecDeque;

fn generate_factorizations(number: u64) -> Vec<Vec<u64>> {
    let mut factorizations = Vec::new();
    let mut stack = VecDeque::from([(number, 2, Vec::new())]);

    while let Some((n, d, path)) = stack.pop_back() {
        let sqrt_n = (n as f64).sqrt() as u64;
        (d..=sqrt_n).filter(|&i| n % i == 0).for_each(|i| {
            let q = n / i;
            let mut new_path = path.clone();
            new_path.push(i);
            stack.push_back((q, i, new_path.clone()));
            if q >= i {
                let mut factorization = new_path.clone();
                factorization.push(q);
                factorizations.push(factorization);
            }
        });
    }
    factorizations
}

fn distinct_permutations<T: Clone + Eq + Ord>(input: &[T]) -> Vec<Vec<T>> {
    let input = input.iter().sorted().cloned().collect_vec();
    let mut result = Vec::new();
    let mut used = vec![false; input.len()];
    let mut current = Vec::with_capacity(input.len());
    permute_unique(&input, &mut used, &mut current, &mut result);
    result
}

fn permute_unique<T: Eq + Ord + Clone>(
    input: &[T],
    used: &mut [bool],
    current: &mut Vec<T>,
    result: &mut Vec<Vec<T>>,
) {
    if current.len() == input.len() {
        result.push(current.clone());
        return;
    }
    for i in 0..input.len() {
        if used[i] || (i > 0 && input[i] == input[i - 1] && !used[i - 1]) {
            continue;
        }
        if !used[i] {
            used[i] = true;
            current.push(input[i].clone());
            permute_unique(input, used, current, result);
            current.pop();
            used[i] = false;
        }
    }
}

pub fn generate_shapes(length: u64) -> Vec<Vec<u64>> {
    let factorizations = generate_factorizations(length);
    factorizations
        .into_iter()
        .flat_map(|f| distinct_permutations(&f))
        .collect()
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct NDShape {
    shape: Vec<usize>,
    strides: Vec<usize>,
    len: usize,
    dimensions: usize,
}

impl NDShape {
    pub fn new(shape: Vec<usize>) -> Self {
        let dimensions = shape.len();
        let strides = Self::calculate_strides(&shape);
        let len = shape.iter().product();
        Self {
            shape,
            strides,
            len,
            dimensions,
        }
    }

    fn calculate_strides(shape: &[usize]) -> Vec<usize> {
        let mut strides = Vec::with_capacity(shape.len());
        let mut current_stride = 1;
        for &dim in shape.iter().rev() {
            strides.push(current_stride);
            current_stride *= dim;
        }
        strides.reverse();
        strides
    }

    pub fn flat_index(&self, indexes: &[usize]) -> usize {
        indexes
            .iter()
            .zip(&self.strides)
            .map(|(&i, &s)| i * s)
            .sum()
    }

    pub fn multi_index(&self, index: usize) -> Vec<usize> {
        let mut remaining = index;
        self.strides
            .iter()
            .map(|&stride| {
                let idx = remaining / stride;
                remaining %= stride;
                idx
            })
            .collect()
    }

    pub fn multi_slices(&self) -> Vec<Vec<Vec<usize>>> {
        (0..self.dimensions)
            .flat_map(|axis| {
                let axis_len = self.shape[axis];
                let ranges = self.shape
                    .iter()
                    .enumerate()
                    .filter_map(|(i, &d)| (i != axis).then(|| 0..d))
                    .collect_vec();
                
                ranges
                    .into_iter()
                    .multi_cartesian_product()
                    .map(move |indices| {
                        (0..axis_len)
                            .map(|m| {
                                let mut cloned = indices.clone();
                                cloned.insert(axis, m);
                                cloned
                            })
                            .collect_vec()
                    })
                    .collect_vec()
            })
            .collect_vec()
    }

    pub fn flat_slices(&self) -> Vec<Vec<usize>> {
        self.multi_slices()
            .into_iter()
            .map(|slice| {
                slice
                    .into_iter()
                    .map(|indices| self.flat_index(&indices))
                    .collect_vec()
            })
            .collect_vec()
    }

    pub fn len(&self) -> usize {
        self.len
    }

    pub fn is_empty(&self) -> bool {
        self.len == 0
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_all_shapes() {
        println!("{:?}", generate_shapes(12));
    }

    #[test]
    fn test_ndshape() {
        let nd_shape = NDShape::new(vec![2, 2]);
        println!("{:?}", nd_shape.flat_slices());
    }
}
